sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/ShardReadersPool.java

```diff
@@ -145,8 +148,15 @@ private void readLoop(ShardRecordsIterator shardRecordsIterator, RateLimitPolicy
           List<KinesisRecord> kinesisRecords = shardRecordsIterator.readNextBatch();
           try {
             for (KinesisRecord kinesisRecord : kinesisRecords) {
-              recordsQueue.put(kinesisRecord);
-              numberOfRecordsInAQueueByShard.get(kinesisRecord.getShardId()).incrementAndGet();
+              while (true) {
+                if (!poolOpened.get()) {
```

Comment:
Why do we need to double-check this?

Reply:
Goal here is to stop as soon as possible to not delay shutting down the reader. The loop to enqueue all new record can take very long, especially using aggregated records. Therefore it's important to check if the pool is still open before every new attempt to offer a record to the queue.